src/middleware/routeHandler.ts

```diff
@@ -0,0 +1,250 @@
+import logger from "@root/logger/logger"
+import convertNeverThrowToPromise from "@root/utils/neverthrow"
+import GitFileSystemService from "@services/db/GitFileSystemService"
+
+const BRANCH_REF = config.get("github.branchRef")
+const simpleGitInstance = simpleGit({
+  maxConcurrentProcesses: MAX_CONCURRENT_GIT_PROCESSES,
+})
+const gitFileSystemService = new GitFileSystemService(simpleGitInstance)
+
+const handleGitFileLock = async (
+  repoName: string,
+  next: (arg0: any) => void
+) => {
+  const result = await gitFileSystemService.hasGitFileLock(repoName, true)
+  if (result.isErr()) {
+    next(result.error)
+    return false
+  }
+  const isGitLocked = result.value
+  if (isGitLocked) {
+    logger.error(`Failed to lock repo ${repoName}: git file system in use.`)
+    next(
+      new LockedError(
+        `Someone else is currently modifying repo ${repoName}. Please try again later.`
+      )
+    )
+    return false
+  }
+  return true
+}
+
+// Used when there are no write API calls to the repo on GitHub
+export const attachReadRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  routeHandler(req, res).catch((err: any) => {
+    next(err)
+  })
+}
+
+// Used when there are write API calls to the repo on GitHub
+export const attachWriteRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  const { siteName } = req.params
+  const { growthbook } = req
+
+  let isGitAvailable = true
+
+  // only check git file lock if the repo is ggs enabled
+  if (growthbook?.getFeatureValue(FEATURE_FLAGS.IS_GGS_ENABLED, false)) {
+    isGitAvailable = await handleGitFileLock(siteName, next)
+  }
+
+  if (!isGitAvailable) return
+
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  await routeHandler(req, res, next).catch(async (err: any) => {
+    await unlock(siteName)
+    next(err)
+  })
+
+  try {
+    await unlock(siteName)
+  } catch (err) {
+    next(err)
+  }
+}
+
+export const attachRollbackRouteHandlerWrapper = (routeHandler: any) => async (
+  req: any,
+  res: any,
+  next: any
+) => {
+  const { userSessionData } = res.locals
+  const { siteName } = req.params
+
+  const { accessToken } = userSessionData
+  const { growthbook } = req
+
+  const shouldUseGitFileSystem = !!growthbook?.getFeatureValue(
+    FEATURE_FLAGS.IS_GGS_ENABLED,
+    false
+  )
+
+  const isGitAvailable = await handleGitFileLock(siteName, next)
+  if (!isGitAvailable) return
+  try {
+    await lock(siteName)
+  } catch (err) {
+    next(err)
+    return
+  }
+
+  let originalStagingCommitSha: any
+  let originalStagingLiteCommitSha: any
+
+  if (shouldUseGitFileSystem) {
+    const results = await Promise.all([
+      gitFileSystemService.getLatestCommitOfBranch(siteName, STAGING_BRANCH),
+      gitFileSystemService.getLatestCommitOfBranch(
+        siteName,
+        STAGING_LITE_BRANCH
+      ),
+    ])
```

Comment:
because this returns result async, which wraps the promise underneath, i'm not entirely sure how `Promise.all` (a native js construct) interacts with this.

`neverthrow` exports `ResultAsync.combine` which does the same thing, but for `ResultAsync` - perhaps we should use that instead?